### PR TITLE
Encoded Column + Null Suppression Framework v1

### PR DESCRIPTION
**Pull request by mjendruk**
Hi!

First of all, I’m sorry for the size of this pull request. I really tried to keep it down to the necessities required to add additional column types and null suppression techniques.

The pull request adds the possibility to add additional column encodings (FOR, Delta, RLE, etc.) using different null suppression techniques as a backend. Null suppression simply means omitting leading zeros in vectors of integers in order to reduce the overall size. Currently there are only two null suppression techniques implemented: our default implementation (byte aligned: 1, 2, and 4 bytes as in `FittedAttributeVector<>`) and SIMD-BP128 (as described in [this paper](https://openproceedings.org/2017/conf/edbt/paper-146.pdf)). There are currently two dictionary column implementations. The old version and the new that is making use of the new null suppression techniques. Sooner or later, the old version will be completely replaced by the new, but this means rewriting the import and export classes, which is a whole pull request in its own.

__The ultimate goal is of course being able to add new compression techniques without ever having to touch the operators. This is already possible with the exception of import and export.__

#### Encoded Columns
 
Column encoding types are listed in `storage/encoded_column/column_encoding_type.hpp`. There is an enum class containing all possible encodings as well as a map with supported data types for each encoding. This ensures that the related template classes are only instantiated with the supported types. All encoded columns, i.e. currently the two dictionary encodings, inherit from `BaseEncodedColumn`. In `storage/encoded_column/utils.hpp` there is a function called `resolve_encoded_column_type` that resolves the type of an encoded column.

Each encoding type has an encoder in `storage/column_encoders`. Columns are encoded via the utility function `encode_column` in `storage/column_encoders/utils.hpp`, which accepts an `EncodingType`. In order to make things easier, there is a template file (`storage/column_encoders/column_encoder.template.hpp`) for adding new column encoders. It’s actually pretty easy. Copy the file, rename the class, implement a templated method and you’re done, almost.

In the future, I’m planning to add column decoders as well. They are going to be used inside of `ReferenceColumnIterable` and will basically have just one method: `std::pair<T, bool> get(const ChunkOffset chunk_offset)`. 

**Edit**: The advantage over adding a method to the column interface is that decoders will be created per context and thus will be able use implement caches.

#### Null Suppression

Null suppression is the physical compression part underlying encoded columns. All relevant files are in `storage/null_suppression`. Each null suppression technique is divided into four conceptual parts: the encoder, which compresses a vector of integers; the vector, an immutable structure holding the encoded data; the decoder, a random-access interface (potentially with caches); and the constant forward iterator. `encode_by_ns_type` in `storage/null_suppression/utils.hpp` encodes a vector with a given type. There are currently only two basic types (as described above). There is a template file (`ns_vector.template.hpp`) for adding new null suppression vectors. All types are listed in `ns_type.hpp`.

#### Previously

export-4

#### Concept

export-6

### ToDo

- [x] I’m going to add tests next (but please go ahead and have a look at the code!)

### Benchmarks

The following benchmark was conducted on the “pella” server. The benchmark executes a table scan on a table with just one chunk, 2 columns of type `int` and 4.000.000 rows. The value range between [0, 10.000]. The table scan has been optimised for all types of dictionary encodings.

```
-----------------------------------------------------------------------------------------
Benchmark                                           Time           CPU Iterations  Factor
-----------------------------------------------------------------------------------------
ValueColumn                                  40136785 ns   39998055 ns         66    1.00
DictionaryColumn                             50725590 ns   50568906 ns         55    1.26
NewDictionaryColumn (FixedSizeByteAligned)   49508383 ns   49354421 ns         57    1.23
NewDictionaryColumn (SimdBp128)              50120835 ns   49971328 ns         57    1.25
```

### CODE REVIEW

**Comment by mrks** [comment]
I will of course be happy to have a look at it, but it might take a while. One question before I get started on that: Do you have any performance numbers for this?

**Comment by mjendruk** [comment]
> I will of course be happy to have a look at it, but it might take a while. One question before I get started on that: Do you have any performance numbers for this?

Everything that’s in the pull request is the basis for the following benchmarks :) But it’s easy to set up a simple benchmark now. I’ll add one.

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+  /**
+   * @return an integral constant implicitly convertible to bool
+   *
+   * Hint: Use decltype(result)::value if you want to use the result
+   *       in a constant expression such as constexpr-if.
+   */
+  template <typename ColumnDataType>
+  auto supports(hana::basic_type<ColumnDataType> data_type) const {
+    return encoding_supports_data_type(Derived::_encoding_type, data_type);
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]:
```➜  cmake-build-release git:(max/null_suppression) ✗ ./hyriseBenchmark --benchmark_filter=TableScan
Run on (8 X 2900 MHz CPU s)
2017-12-06 11:46:53
[1]    47214 killed     ./hyriseBenchmark --benchmark_filter=TableScan```

[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: Works with the latest commit.

**Comment by Bensk1** [comment]
The introduction you wrote mitigates the size of the pull request. Some general questions/remarks: 

- Should we really call it new dictionary? Even given that it is only a temporary name, I do not like it too much
- I see the point that some parts of the literature call the technique `null suppression`, but as stated yesterday during the fortnightly, most people first think of NULL values and I believe this naming might lead to confusion when `null suppression` and `is_null[able]()` are used next to each other in the same file
- "Sooner or later, the old version will be completely replaced by the new" This means we will remove the old class from Hyrise, but the old technique of simple dictionary encoding will still be part of Hyrise, right?

**Comment by mjendruk** [comment]
> The introduction you wrote mitigates the size of the pull request. Some general questions/remarks: 

> - Should we really call it new dictionary? Even given that it is only a temporary name, I do not like it too much

Do you have any suggestions?

> - I see the point that some parts of the literature call the technique `null suppression`, but as stated yesterday during the fortnightly, most people first think of NULL values and I believe this naming might lead to confusion when `null suppression` and `is_null[able]()` are used next to each other in the same file

We could call it zero suppression?

> - "Sooner or later, the old version will be completely replaced by the new" This means we will remove the old class from Hyrise, but the old technique of simple dictionary encoding will still be part of Hyrise, right?

Yes, the interface has slightly changed (`attribute_vector()` returns an `BaseNsVector` now) and some classes haven’t been updated yet. I’m going to do this later as it will require some bigger additions in the import and export classes.

**Comment by Bouncner** [comment]
While I prefer null suppression as it is the mostly used term, I totally see the point you guys brought up.
+1 for 'zero surpression'

**Comment by TwoLaid** [comment]
> decreased by 3.7 compared to master and is now 86.4%

hmmmm.....

**Comment by mjendruk** [comment]
>> Coverage: decreased by 3.7 compared to master and is now 86.4%

> hmmmm.....

Please tell me what to do. A lot of the ”untested” code in fact tested, but not recognized.

**Comment by Bensk1** [comment]
>> - Should we really call it new dictionary? Even given that it is only a temporary name, I do not like it too much

> Do you have any suggestions?

Not really. Maybe `ZeroSuppressionDictionary` or simply `OldDictionary` and `Dictionary`?

**Comment by mjendruk** [comment]
>>> Should we really call it new dictionary? Even given that it is only a temporary name, I do not like it too much

>> Do you have any suggestions?

> Not really. Maybe `ZeroSuppressionDictionary` or simply `OldDictionary` and `Dictionary`?

We could call the old one deprecated (and also mark it with `[[deprecated]]`), i.e., `DeprecatedDictionaryColumn`?

**Comment by Bensk1** [comment]
Sounds good to me.

**Comment by mjendruk** [comment]
> We could call the old one deprecated (and also mark it with [[deprecated]]), i.e., DeprecatedDictionaryColumn?

Done. I didn’t add the `deprecated` attribute because it fills the console with thousands of warnings.

**Comment by TwoLaid** [comment]
For test coverage:

* For `SimdBp128Packing` (and unpacking) only the bitcases 11-16 are covered.
* `NsVector` methods like `NsVector::decode` are not used. I put prints in there and didn't see them while executing the tests
* Same with `FixedSizeByteAlignedVector`, because none of the `_on_*` methods are called by the `NsVector`
* Same with `SimdBp128Iterator`, because `_on_cbegin` is never called

I can't really put my finger on why all this code is unused at the moment, but I'm sure you'll figure it out :)

File: src/benchmark/benchmark_basic_fixture.cpp [hunk 2]
```diff
@@ -22,10 +22,14 @@ void BenchmarkBasicFixture::SetUp(::benchmark::State& state) {
```
Comment (mjendruk) [inline src/benchmark/benchmark_basic_fixture.cpp]: > For test coverage:
> 
> * For `SimdBp128Packing` (and unpacking) only the bitcases 11-16 are covered.

Yes that’s true, but it’s just a tiny part of what is supposedly not covered.

> * `NsVector` methods like `NsVector::decode` are not used. I put prints in there and didn't see them while executing the tests

Yes that’s true, too. I added another tests for this method.

> * Same with `FixedSizeByteAlignedVector`, because none of the `_on_*` methods are called by the `NsVector`

This is not true. Have a look at `null_suppression_test.cpp`. The iterator methods and `create_base_decoder` are in fact called.

> * Same with `SimdBp128Iterator`, because `_on_cbegin` is never called

I tested this via print debugging and it gets in fact called.

> I can't really put my finger on why all this code is unused at the moment, but I'm sure you'll figure it out :)

Furthermore some of the new untested code, I didn’t even touch (`abstract_lqp_node.cpp`).

@mrks Any ideas how we can solve this?

**Comment by mjendruk** [comment]
> I see the point that some parts of the literature call the technique null suppression, but as stated yesterday during the fortnightly, most people first think of NULL values and I believe this naming might lead to confusion when null suppression and is_null[able]() are used next to each other in the same file

Rename to zero suppression ✓

File: src/benchmark/benchmark_basic_fixture.cpp (hunk 2 above)
Comment (TwoLaid) [reply src/benchmark/benchmark_basic_fixture.cpp]: I just realized that you are 8 commits behind `master`, so you are probably missing my commits for all the LQP tests

Comment (TwoLaid) [reply src/benchmark/benchmark_basic_fixture.cpp]: > Coverage — decreased by 4.3 compared to master and is now 85.7%

I took the liberty of merging the `master` but now it's even worse - oops.

> Yes that’s true, but it’s just a tiny part of what is supposedly not covered.

Actually, for `simd_bp128_packing.cpp` alone, the non-covered bitcases make up about 160 lines of code alone. I think that's pretty significant.

Other than that, after skimming the pycobertura report it seems like it's just a lot of new Dictionary functionalities that are not covered.

Comment (mjendruk) [reply src/benchmark/benchmark_basic_fixture.cpp]: > I took the liberty of merging the master but now it's even worse - oops.

Thanks …

Comment (mjendruk) [reply src/benchmark/benchmark_basic_fixture.cpp]: @TwoLaid okay the test coverage decreases now only by ~1.3%. The true coverage is higher, because a lot of templated method are in fact covered. We need to fix somehow, but maybe not in this pr.

**Comment by mjendruk** [comment]
@mrks @mrzzzrm Everything is done from my side. Could you have a look?

**Comment by mrks** [comment]
Sorry for the delay. I am still busy with getting that paper ready for submission, but I am fully committed to working through this before I leave for Christmas.

File: src/lib/operators/table_scan/single_column_table_scan_impl.cpp [hunk 3]
```diff
@@ -61,13 +78,79 @@ void SingleColumnTableScanImpl::handle_value_column(const BaseValueColumn& base_
   });
 }
 
-void SingleColumnTableScanImpl::handle_dictionary_column(const BaseDictionaryColumn& base_column,
+void SingleColumnTableScanImpl::handle_dictionary_column(const BaseDeprecatedDictionaryColumn& base_column,
                                                          std::shared_ptr<ColumnVisitableContext> base_context) {
+  _handle_dictionary_column(base_column, base_context);
+
+  // Implementation of an unoptimized scan
```
Comment (mrks) [inline src/lib/operators/table_scan/single_column_table_scan_impl.cpp]: Any way we can have this as actual, compiled code? Otherwise, people will change the code that is compiled and this comment will quickly become outdated.

File: src/lib/operators/table_scan/single_column_table_scan_impl.cpp [hunk 4]
```diff
@@ -61,13 +78,79 @@ void SingleColumnTableScanImpl::handle_value_column(const BaseValueColumn& base_
...
+  //         _binary_scan(comparator, left_it, left_end, right_it, chunk_id, matches_out);
+  //       });
+  //     });
+  //   });
+  // });
+}
+
+void SingleColumnTableScanImpl::handle_encoded_column(const BaseEncodedColumn& base_column,
+                                                      std::shared_ptr<ColumnVisitableContext> base_context) {
+  if (base_column.encoding_type() == EncodingType::Dictionary) {
```
Comment (mrks) [inline src/lib/operators/table_scan/single_column_table_scan_impl.cpp]: This needs a comment.

File: src/lib/operators/table_scan/single_column_table_scan_impl.cpp [hunk 5]
```diff
@@ -61,13 +78,79 @@ void SingleColumnTableScanImpl::handle_value_column(const BaseValueColumn& base_
...
+
+    resolve_encoded_column_type<Type>(base_column, [&](const auto& left_column) {
+      auto left_column_iterable = create_iterable_from_column(left_column);
+
+      auto right_value_iterable = ConstantValueIterable<Type>{_right_value};
+
+      left_column_iterable.with_iterators(mapped_chunk_offsets.get(), [&](auto left_it, auto left_end) {
+        right_value_iterable.with_iterators([&](auto right_it, auto right_end) {
+          with_comparator(_scan_type, [&](auto comparator) {
+            _binary_scan(comparator, left_it, left_end, right_it, chunk_id, matches_out);
```
Comment (mrks) [inline src/lib/operators/table_scan/single_column_table_scan_impl.cpp]: This looks like code duplication from the value column?

File: src/lib/storage/base_column.hpp [hunk 6]
```diff
@@ -32,6 +32,9 @@ class BaseColumn : private Noncopyable {
   // returns the number of values
   virtual size_t size() const = 0;
 
+  // Returns the physical size of the column
+  virtual size_t data_size() const = 0;
```
Comment (mrks) [inline src/lib/storage/base_column.hpp]: Does this take nasty things like strings into account?

File: src/lib/storage/base_dictionary_column.hpp [hunk 7]
```diff
@@ -1,27 +0,0 @@
-#pragma once
```
Comment (mrks) [inline src/lib/storage/base_dictionary_column.hpp]: Since this will change the interface, could you please take the time to sit down with the seminar groups that are using it? This is mostly the data types group, but the join and pruning groups might be affected too. Feel free to do this in your Hiwi time.

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+#include "all_type_variant.hpp"
+#include "resolve_type.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+namespace hana = boost::hana;
+
+/**
+ * @brief Base class of all column encoders
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Do you have an extensive comment on how everything works somewhere? This could use a pointer to that comment (or a wiki entry)

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+ *
+ * Use the column_encoder.template.hpp to add new implementations!
+ */
+class BaseColumnEncoder {
+ public:
+  /**
+   * @brief Returns true if the encoder supports the given data type.
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: What would be a case where it is unsupported?

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  virtual bool supports(DataType data_type) const = 0;
+
+  /**
+   * @brief Encodes a value column with the given data type.
+   *
+   * @return encoded column if data type is supported else nullptr
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Do you rely on the nullptr part? I think callers should call `supports` before and `encode` should fail if it is not supported

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+template <typename Derived>
+class ColumnEncoder : public BaseColumnEncoder {
+ public:
+  /**
+   * @defgroup Virtual interface implementation
+   * @{
+   */
+  bool supports(DataType data_type) const final {
+    auto result = bool{};
+    resolve_data_type(data_type, [&](auto type_obj) { result = this->supports(type_obj); });
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: comment?

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * @brief Encodes a value column with the given data type.
+   *
+   * @return encoded column if data type is supported else nullptr
+   */
+  virtual std::shared_ptr<BaseColumn> encode(DataType data_type, const std::shared_ptr<BaseValueColumn>& column) = 0;
+};
+
+template <typename Derived>
+class ColumnEncoder : public BaseColumnEncoder {
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Why is this in base_column_encoder.hpp?

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+   * @defgroup Virtual interface implementation
+   * @{
+   */
+  bool supports(DataType data_type) const final {
+    auto result = bool{};
+    resolve_data_type(data_type, [&](auto type_obj) { result = this->supports(type_obj); });
+    return result;
+  }
+
+  std::shared_ptr<BaseColumn> encode(DataType data_type, const std::shared_ptr<BaseValueColumn>& column) final {
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Comment? Something like "dispatches to the appropriate instantiation of "encode"?

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * @defgroup Non-virtual interface
+   * @{
+   */
+
+  /**
+   * @return an integral constant implicitly convertible to bool
+   *
+   * Hint: Use decltype(result)::value if you want to use the result
+   *       in a constant expression such as constexpr-if.
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Why is this necessary? Why can't you return bool?

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+   * @brief Encodes a value column with the given data type.
+   *
+   * Compiles only for supported data types.
+   */
+  template <typename ColumnDataType>
+  std::shared_ptr<BaseColumn> encode(hana::basic_type<ColumnDataType> data_type,
+                                     const std::shared_ptr<BaseValueColumn>& value_column) {
+    static_assert(decltype(supports(data_type))::value);
+
+    return _self()._encode(std::static_pointer_cast<ValueColumn<ColumnDataType>>(value_column));
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Do we have three different "encode" methods, one with an underscore? Can we find more specific names for those?

File: src/lib/storage/column_encoders/base_column_encoder.hpp [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+  }
+  /**@}*/
+
+ private:
+  Derived& _self() { return static_cast<Derived&>(*this); }
```
Comment (mrks) [inline src/lib/storage/column_encoders/base_column_encoder.hpp]: Maybe a comment about CRTP somewhere? I don't think we can expect people to know about it.

File: src/lib/storage/column_encoders/column_encoder.template.hpp [hunk 17]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Column Encoder Template
+ *
+ * - Copy this file for new column encoders
+ * - Add new encoder to encoders.hpp
+ */
+#pragma once
```
Comment (mrks) [inline src/lib/storage/column_encoders/column_encoder.template.hpp]: Is this compiled and tested somewhere? Code that is not part of the program *will* become outdated.

File: src/lib/storage/column_encoders/column_encoder.template.hpp [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "base_column_encoder.hpp"
+
+#include "storage/value_column.hpp"
+#include "types.hpp"
+#include "utils/enum_constant.hpp"
+
+namespace opossum {
+
+// TODO(you): Rename class
```
Comment (mrks) [inline src/lib/storage/column_encoders/column_encoder.template.hpp]: Not a fan of using "TODO" here, because it would show up when people are looking for todos in the project.

File: src/lib/storage/column_encoders/deprecated_dictionary_encoder.hpp [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+#include "utils/enum_constant.hpp"
+
+namespace opossum {
+
+class DeprecatedDictionaryEncoder : public ColumnEncoder<DeprecatedDictionaryEncoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::DeprecatedDictionary>;
+
+  template <typename T>
+  std::shared_ptr<BaseColumn> _encode(const std::shared_ptr<ValueColumn<T>>& value_column) {
```
Comment (mrks) [inline src/lib/storage/column_encoders/deprecated_dictionary_encoder.hpp]: Assuming that nothing changed here. Please correct me if that's wrong.

File: src/lib/storage/column_encoders/dictionary_encoder.hpp [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+      // Erase null values
+      dictionary.erase(erase_from_here_it, dictionary.end());
+    }
+
+    std::sort(dictionary.begin(), dictionary.end());
+    dictionary.erase(std::unique(dictionary.begin(), dictionary.end()), dictionary.end());
+    dictionary.shrink_to_fit();
+
+    auto attribute_vector = pmr_vector<uint32_t>{values.get_allocator()};
```
Comment (mrks) [inline src/lib/storage/column_encoders/dictionary_encoder.hpp]: So we are first creating a standard-size vector and will compress it later? This should be documented somewhere.

File: src/lib/storage/column_encoders/dictionary_encoder.hpp [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+ private:
+  template <typename T>
+  static ValueID _get_value_id(const pmr_vector<T>& dictionary, const T& value) {
+    return static_cast<ValueID>(
+        std::distance(dictionary.cbegin(), std::lower_bound(dictionary.cbegin(), dictionary.cend(), value)));
+  }
+
+  ZsType get_fixed_size_byte_aligned_encoding(size_t unique_values_count) {
```
Comment (mrks) [inline src/lib/storage/column_encoders/dictionary_encoder.hpp]: comment?

File: src/lib/storage/column_encoders/utils.hpp [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+namespace opossum {
+
+class BaseColumn;
+class BaseColumnEncoder;
+class BaseValueColumn;
+
+/**
+ * @brief Creates an encoder by encoding type
+ */
+std::unique_ptr<BaseColumnEncoder> create_encoder(EncodingType encoding_type);
```
Comment (mrks) [inline src/lib/storage/column_encoders/utils.hpp]: Do we have to put this into the global namespace, especially if there is only one user? How about `BaseColumnEncoder::create_encoder`?

File: src/lib/storage/column_encoders/utils.hpp [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * @brief Encodes a value column by the given encoding method
+ *
+ * @return encoded column if data type is supported else nullptr
+ */
+std::shared_ptr<BaseColumn> encode_column(EncodingType encoding_type, DataType data_type,
```
Comment (mrks) [inline src/lib/storage/column_encoders/utils.hpp]: see above

File: src/lib/storage/encoded_columns/base_dictionary_column.hpp [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+
+  /**
+   * @brief The size of the dictionary
+   */
+  virtual size_t unique_values_count() const = 0;
+
+  virtual std::shared_ptr<const BaseZeroSuppressionVector> attribute_vector() const = 0;
+
+  /**
+   * @brief Returns encoding specific null value ID
```
Comment (mrks) [inline src/lib/storage/encoded_columns/base_dictionary_column.hpp]: Why do we have encoding-specific null ids?

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 12 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: As in `iterables.hpp`, I put the base class and its CRTP sub-class into the same file. Would be `column_encoder.hpp` better?

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 14 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: `encoder.supports(data_type);` is not a constant expression unless `encoder` is constexpr, which in most cases it isn’t. By returning a integral constant, it is possible to retrieve the result type as a constant expression and therefore use it in for example constexpr-ifs. Did this help?

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 15 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: Yes, the `_encode` version should be `_on_encode`. This is the naming I used in all other cases where I use CRTP. With regards to the other two versions (and support methods), the difference is that the virtual methods accept `DataType` the enum as the data type specification while the non-virtual methods require true types as template parameters. The virtual methods call the non-virtual methods, but in order to do so, they need to resolve the data type.

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 16 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: There is one in `iterables.hpp`. Moving the information into the wiki in referring to it seems a good idea. What do you think?

File: src/lib/storage/column_encoders/column_encoder.template.hpp (hunk 17 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/column_encoder.template.hpp]: This file should be updated if the interface changes. They is a similar already existing file:`test/gtest_case_template.cpp`. You are of course right, it is easy to forgot updating this file. Shall I remove it?

File: src/lib/storage/column_encoders/column_encoder.template.hpp (hunk 18 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/column_encoder.template.hpp]: Yes

File: src/lib/storage/column_encoders/deprecated_dictionary_encoder.hpp (hunk 19 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/deprecated_dictionary_encoder.hpp]: Yes, nothing changed. I just moved the code.

File: src/lib/storage/column_encoders/dictionary_encoder.hpp (hunk 20 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/dictionary_encoder.hpp]: Okay

File: src/lib/storage/column_encoders/utils.hpp (hunk 22 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/utils.hpp]: Mhm, should a base column know about all of its sub classes?

File: src/lib/storage/encoded_columns/base_dictionary_column.hpp (hunk 24 above)
Comment (mjendruk) [reply src/lib/storage/encoded_columns/base_dictionary_column.hpp]: We already had them before, but the virtual interface of `BaseAttributeVector` hid this from the user with the cost of having a branch in each `get(ValueID)`. The way we did in the old implementation is however not generalisable. SIMD-BP128 for example has a variable bit size so the maximum possible value depends on the segment. So the easiest solution was to set the null value ID to dictionary size + 1. This also avoids also the unnecessary conversion inside of `BaseAttributeVector`.

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 8 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: I’ll add one.

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 9 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: FOR on strings for example.

File: src/lib/operators/export_binary.cpp [hunk 25]
```diff
@@ -237,6 +237,12 @@ void ExportBinary::ExportBinaryVisitor<T>::handle_dictionary_column(
   _export_attribute_vector(context->ofstream, *column.attribute_vector());
 }
 
+template <typename T>
+void ExportBinary::ExportBinaryVisitor<T>::handle_encoded_column(const BaseEncodedColumn& base_column,
+                                                                 std::shared_ptr<ColumnVisitableContext> base_context) {
+  Fail("Binary export not implemented for encoded columns.");
```
[the remark this answers is not in the document]
Comment (mjendruk) [reply src/lib/operators/export_binary.cpp]: Yes!

File: src/lib/operators/table_scan/single_column_table_scan_impl.cpp (hunk 5 above)
Comment (mjendruk) [reply src/lib/operators/table_scan/single_column_table_scan_impl.cpp]: Yes, I updated the TableScan. This part is gone now.

File: src/lib/operators/table_scan/single_column_table_scan_impl.cpp (hunk 4 above)
Comment (mjendruk) [reply src/lib/operators/table_scan/single_column_table_scan_impl.cpp]: I refactored it.

File: src/lib/operators/table_scan/single_column_table_scan_impl.cpp (hunk 3 above)
Comment (mjendruk) [reply src/lib/operators/table_scan/single_column_table_scan_impl.cpp]: I removed it.

File: src/lib/storage/base_column.hpp (hunk 6 above)
Comment (mjendruk) [reply src/lib/storage/base_column.hpp]: No, thanks for reminding me of strings. I have removed these methods for now.

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 14 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: I did not know what I explained before I implemented this method to be honest. I have added a short comment.

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 13 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: done

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 11 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: I’m not sure what I should write. `resolve_data_type` is already explained and used in a lot of places and `supports` is explained a bit further down.

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 10 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: Done

File: src/lib/storage/column_encoders/base_column_encoder.hpp (hunk 8 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/base_column_encoder.hpp]: Edit: I think we have to talk about this together. I have worked too long on this code to know what should be explained and what not.

File: src/lib/storage/column_encoders/utils.hpp (hunk 22 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/utils.hpp]: I don’t know. I don’t see a problem here. This file is only included in a cpp-file and isn’t in the global namespace except in this one file that includes it.

File: src/lib/storage/column_encoders/dictionary_encoder.hpp (hunk 20 above)
Comment (mjendruk) [reply src/lib/storage/column_encoders/dictionary_encoder.hpp]: done

**Comment by mjendruk** [comment]
> In general I think that this really raises the entry barrier. For that, there are too few comments. Right now, I don't see how we could get a new round of seminar students started on this.

Okay. I have addressed or answered to all your comments. To me, a lot of it seems like applying what I already did with the iterables. As I wrote in one of the comments, I’ve spent too much time with this code to know which parts are clear and which are not. We should talk about this in person. The diagram at the top and the description at the top should help to get an overview. Maybe this should become a wiki page.

@mrks
